### PR TITLE
8251946: ObservableList.setAll does not conform to specification

### PR DESCRIPTION
**Pull request by TheMrMilchmann**
Hi, this PR fixes [JDK-8251946](https://bugs.openjdk.java.net/browse/JDK-8251946) computing whether the list was actually modified instead of just returning `true`. The list was modified if
1. it was not empty (modified by calling `#clear()`), or if
2. it was modified as result of the `#addAll()` call.

If you want any test coverage for this please let me know.

I reported the issue a couple of days ago via web formula and waited for the confirmation to open this PR but now I see that @kevinrushforth (sorry for pinging) is already assigned to the JBS issue. I'm not too familiar with how you handle JBS issues or more specifically whether assigning is used to indicate that the issue is in the assignee's "domain", or that the assignee is already working on it. In the latter case, please feel free to close this PR. (My OCA submission is still pending anyway.)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8251946](https://bugs.openjdk.java.net/browse/JDK-8251946): ObservableList.setAll does not conform to specification

### Reviewers
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/284/head:pull/284`
`$ git checkout pull/284`

### CODE REVIEW

**Comment by kevinrushforth** [comment]
@TheMrMilchmann I am not actively working on that bug, so you can proceed with this PR. I will review it when it is ready.

Once you have submitted the OCA, please add a comment to this PR with the `/signed` command (and nothing else in the comment). After your OCA is approved and recorded, the Skara bot will mark it as ready for review.

Please do add a unit test for this. You can find existing collections tests [here](https://github.com/openjdk/jfx/tree/master/modules/javafx.base/src/test/java/test/javafx/collections).

**Comment by kevinrushforth** [comment]
One overall comment while we are waiting for your OCA to be approved.

I don't think the complexity of this proposed fix to `setAll` is warranted. I would prefer a simpler fix that returns `false` if both the current list and the new list are empty, and `true` otherwise. This method is essentially a convenience method that does:

List::clear
List::addAll(...)

Given this, it seems reasonable for `setAll` to return true if either `clear` or `addAll` would modify the list.

If there is a good justification for handling the corner case of calling `setAll` with the same list of elements in exactly the same order (and I am not sure that there is), then a better approach might be to do the check before actually modifying the list, returning early if the new list and the current list were identical. In that case, the list really would be unmodified and no listeners would be notified.

**Comment by robilad** [comment]
Unfortunately, I don't see @TheMrMilchmann  OCA in the approval queue. Could you please (re)send it to Oracle-ca_us@oracle.com? Thanks!

**Comment by TheMrMilchmann** [comment]
> I don't think the complexity of this proposed fix to `setAll` is warranted. I would prefer a simpler fix that returns `false` if both the current list and the new list are empty, and `true` otherwise. 

@kevinrushforth Thanks for your feedback! I have just pushed a commit that greatly simplifies `setAll` by returning early if both lists are empty and continuing with `clear`, `addAll` otherwise.

> If there is a good justification for handling the corner case of calling `setAll` with the same list of elements in exactly the same order (and I am not sure that there is), then a better approach might be to do the check before actually modifying the list, returning early if the new list and the current list were identical.

Fair point. I'm afraid I don't have any justification other than that, in my opinion, this is a violation of the method's contract (which is debatable). I suppose it's best to leave it as-is for this case for now.

---

> Unfortunately, I don't see @TheMrMilchmann OCA in the approval queue. Could you please (re)send it to Oracle-ca_us@oracle.com? Thanks!

@robilad That's strange. I have sent a mail on Aug 13 and received no indication that it didn't go through (though neither did I receive an automated confirmation that it arrived - if such a thing is set up). Anyway, I just resent it.

**Comment by robilad** [comment]
Thanks @TheMrMilchmann and apologies for the delay!

File: modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.ListIterator;
+import java.util.Objects;
```
Comment (arapte) [inline modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:32]: This import seems unused, please remove.

File: modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java [hunk 2]
```diff
@@ -88,14 +89,15 @@ public ModifiableObservableListBase() {
 
     @Override
     public boolean setAll(Collection<? extends E> col) {
+        if (isEmpty() && col.isEmpty()) return false;
         beginChange();
         try {
             clear();
             addAll(col);
+            return true;
```
Comment (arapte) [inline modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:97]: I think following code would be more suitable here, 
```
boolean res = super.addAll(c);
return res;
```
This code is already used in two `addAll()` methods of this class.

**Review by arapte — CHANGES_REQUESTED** [review]
Change seems good, suggested minor changes.

File: modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java (hunk 1 above)
Comment (TheMrMilchmann) [reply modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java]: Indeed. Removed it.

File: modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java (hunk 2 above)
Comment (TheMrMilchmann) [reply modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java]: Makes sense to me. I changed it accordingly.

Comment (kevinrushforth) [reply modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java]: I don't think this change is correct.  `setAll(Collection)` should return true if the list is modified. As discussed in an [earlier comment](#issuecomment-684117392) this means returning true if either the existing Collection or the new Collection is non-empty.

Comment (TheMrMilchmann) [reply modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java:96]: I'm sorry, you're absolutely right. I didn't spend enough time thinking about the change and assumed it was fine since the tests passed. Turns out the test didn't cover it. I reverted that part of the change and added a test-case.

File: modules/javafx.base/src/test/java/test/javafx/collections/ObservableListTest.java [hunk 3]
```diff
@@ -255,6 +255,26 @@ public void testSet() {
...
+    public void testSetAll() {
+        useListData("one", "two", "three");
+        boolean r = list.setAll("one");
+        assertTrue(r);
+
+        r = list.setAll("one", "four", "five");
+        assertTrue(r);
+
+        r = list.setAll();
+        assertTrue(r);
```
Comment (kevinrushforth) [inline modules/javafx.base/src/test/java/test/javafx/collections/ObservableListTest.java:268]: Can you also test that calling `setAll` when the list is currently empty returns true? Repeating one of the earlier checks should work:

```
        r = list.setAll("one");
        assertTrue(r);
```

**Review by kevinrushforth — COMMENTED** [review]
The code change looks good. I have one suggestion for the test.

File: modules/javafx.base/src/test/java/test/javafx/collections/ObservableListTest.java (hunk 3 above)
Comment (TheMrMilchmann) [reply modules/javafx.base/src/test/java/test/javafx/collections/ObservableListTest.java:268]: Sure. Added in #284 5de1a51fd3a8d744acf0b831ab4f5a0441e417d7.

**Review by kevinrushforth — APPROVED** [review]

**Review by arapte — APPROVED** [review]
